[4단계 미션] 임규영 제출합니다.

gxuoo opened the pull request:
안녕하세요 김민석 리뷰어님!

3단계 미션에 이어 4단계 미션에 대한 PR 보내드립니다! 4단계 미션은 `Header`의 레스토랑 추가 버튼을 클릭하면 레스토랑 추가 `form`이 `Modal`로 뜨도록 구현하고, 값을 입력한 후에 '추가하기' 버튼을 클릭하면 레스토랑 목록에 추가되도록 구현하는 것이 목표였습니다. 또한 추가적으로 재사용 가능한 Modal 컴포넌트를 만들어 `RestaurantDetailModal`, `RestaurantAddModal`을 수정해보는 것이 최종적인 미션이었습니다.

## 🗂️ 프로젝트 구조 (Directory Structure)
<img width="959" alt="Step4 Structure" src="https://github.com/user-attachments/assets/4db181ec-4cdc-4ff4-a522-d981ef459a52" />

## 🎆 View Viedo
https://github.com/user-attachments/assets/11c1b609-689f-455c-b5c2-45f2a9459e16


## 컴포넌트 상세 설명
### App.jsx
- 레스토랑 배열을 상테 변수로 관리하기 위해 `restaurantList` 변수를 추가, 각 `Modal`의 인덱스를 추가하여 원하는 `Modal`을 출력하기 위해 `activeIndex` 변수 추가하여 `RestaurantContainer`, `AsideContainer`에 필요한 변수들을 props로 내렸습니다.

### Header.jsx
- `onClick` 이벤트를 통해 `modalState`와 `activeIndex` 값을 변경하여 `Modal`의 open & close, `AddModal`을 활성화하도록 `index`를 설정하는 코드를 추가했습니다.

### Modal.jsx
- `AddModal`과 `DetailModal` 2가지의 겹치는 태그들을 모아 `modal-container` 클래스를 가진 <div> 태그까지 동일하고, `title`에 따라 추가되는 태그 설정, 내용을 담을 수 있는 `children` 속성, 마지막 버튼을 관리해주는 `showClosebutton`, `button-container`의 클래스를 가지는 <div> 태그를 포함하는 컴포넌트 입니다.

### RestaurantAddModal.jsx & RestaurantDetailModal.jsx
- `Modal.jsx`를 import하여 컴포넌트를 재사용하고 더 필요한 부분은 `Modal` 컴포넌트의 `children`으로 구현했습니다.
- `RestaurantAddModal`에는 `form`을 제출했을 때 `handleSubmit` 함수를 호출하여 레스토랑 리스트에 추가하고자 하는 레스토랑 정보를 저장하는 코드를 추가적으로 구현했습니다. 
- 레스토랑이 추가되는 방식은 `restaurant.js`에 선언된 객체 배열에 새로운 객체를 추가하는 방식으로 구현했습니다. 또한 `AddModal`이 보여질 때 `backdrop`을 클릭하면 다시 레스토랑 리스트를 보여주는 방식으로 구현했습니다.

## ❗️중점적으로 봐주셨으면 하는 점과 구현하면서 느낀 점
- 사실 이번 미션을 진행하면서 useState 변수를 너무 많이 선언하지 않았나...?? 라는 생각이 상당히 지배적이었습니다. 여러 차례 혼자 고민을 해봤지만 제가 가지고 있는 지식 한에서는 이정도가 최선이었습니다...😅 따라서 민석님이 보시기에 컴포넌트 구조를 보셨을 때 불필요한 상태 관리가 있다면 수정하도록 하겠습니다. 

- 컴포넌트의 기능 분리는 Header의 경우에는 1가지 컴포넌트 밖에 존재하지 않기에 따로 컨테이너 개념을 도입하지 않아도 될 것 같다고 생각하였고, Main(Restaurant), Aside의 경우 여러 가지 컴포넌트가 존재하기에 컨테이너 개념을 도입하여 해당 컨테이너 내부에서 관리할 수 있는 것들은 최상위 컨테이너까지 올라가지 않아도 생각하여 그렇게 구현했습니다.

이상, 4단계 미션 PR 마치겠습니다. 민석님 편하신 시간대에 천천히 리뷰 주시면 빠르게 리뷰 반영하여 수정하도록 하겠습니다 감사합니다!!

gxuoo committed:
feat: step1 컴포넌트 기초 틀 구현

가장 큰 Container라는 틀을 통해 Head, Main, Aside 3가지로 구분 지었고, 각각의 해당되는 css 파일을 구분했습니다.

gxuoo committed:
feat: Header 컴포넌트 구현

HeadContainer에 들어가는 Header 컴포넌트 구현

gxuoo committed:
feat: 레스토랑 필터, 리스트, 값 컴포넌트 구현

MainContainer에 들어갈 레스토랑 카테고리에 따른 필터와 레스토랑 리스트, 리스트에 들어갈 각각의 레스토랑 값에 대한 컴포넌트를 생성했습니다. 그리고 이에 따른 png 파일의 저장 위치를 React 폴더 구조에 알맞게 변경해보았습니다.

gxuoo committed:
feat: Aside 컴포넌트 구현

레스토랑의 정보를 띄워주는 컴포넌트, 레스토랑을 추가할 수 있는 컴포넌트를 구현했습니다.

gxuoo committed:
fix: 파일명 수정

gxuoo committed:
refactor: 파일명 수정, .png 위치 수정

gxuoo committed:
feat: module.css 적용 및 카테고리 설정에 따른 필터된 음식점 목록 출력

전역 App.css 와 Head, Main, Aside 각 컴포넌트에서 사용되는 css에 대해 코드를 수정하였고, 카테고리 설정에 따른 필터링된 음식점 목록을 보여주는 기능을 추가했습니다.

gxuoo committed:
fix: 리뷰 반영 수정

네이밍 변경, 불필요한 Fragment 제거, 카테고리 컴포넌트 추가, map 함수 key 변경

gxuoo committed:
Merge branch 'gxuoo' into step2

gxuoo committed:
fix: 2차 리뷰 반영 수정

불필요한 Container 컴포넌트 제거, 컴포넌트 외부에 상수 선언, 네이밍 변경

gxuoo committed:
refactor: 2단계 리뷰 반영 리팩토링

사용하지 않는 파일들 삭제

gxuoo committed:
feat: RestaurantDetailModal 구현

레스토랑을 클릭하면 상세 정보를 띄워주는 Modal 구현

gxuoo committed:
Merge remote-tracking branch 'upstream/gxuoo' into step3

gxuoo committed:
refacator: 사용하지 않는 파일 삭제

gxuoo committed:
feat: RestaurantAddModal 구현

레스토랑 추가하는 기능을 구현하였고 재사용 가능한 Modal 컴포넌트를 사용하여 AddModal, DetailModal에 적용시켰습니다.

shackstack commented on src/components/aside/modal/Modal.jsx:
```diff
@@ -0,0 +1,22 @@
+                {title && <h2 className={`${styles["modal-title"]} text-title`}>{title}</h2>}
+                <form onSubmit={onSubmit || ((e) => e.preventDefault())}>
+                    {children}
+                    {showCloseButton && (
```
CloseButton은 닫기 뿐만아니라 추가하기 역할까지 하고있습니다. 적절한 네이밍일까요?

gxuoo replied:
지금 다시보니 네이밍에 문제가 많군요.. 수정하겠습니다.

shackstack commented on src/components/category/RestaurantCategory.jsx:
```diff
@@ -1,41 +1,41 @@
-            name: "기타"
-        }
-    ]
+const categoryList = [
```
컴포넌트 밖으로 잘 분리하셨군요.👍
다만 상수에 대해서 upper case로 규칙을 잡으셨다면 일관성있게 맞추어 주세요.

shackstack commented on src/App.jsx:
```diff
@@ -1,14 +1,32 @@
 
 function App() {
+  const [modalState, setModalState] = useState(false);
+  const [activeIndex, setActiveIndex] = useState(0);
```
원하는 모달의 인덱스를 관리하는 상태라면, `activeModalIndex`가 더 읽히기 쉬울 것 같습니다.

gxuoo replied:
변경했습니다!

shackstack commented on src/App.jsx:
```diff
@@ -1,14 +1,32 @@
+      <RestaurantContainer setModalState={setModalState} 
+                           setRestaurantValue={setRestaurantValue}
+                           restaurantList={restaurantList} />
+      {modalState && <AsideContainer setModalState={setModalState} 
```
AsideContainer가 모달상태에 따라 조건부렌더링 된다면, 그냥 모달 컴포넌트와 다를 바가 없어보입니다.

shackstack reviewed (commented):
### 컴포넌트 구조 관련 문제점
현재 코드베이스는 불필요하게 세분화된 컴포넌트 구조를 가지고 있습니다. 이로 인해 다음과 같은 문제가 발생하고 있습니다:
#### 과도한 Props Drilling:
App.jsx에서 시작된 상태와 함수들이 AsideContainer → RestaurantModal → RestaurantDetailModal/RestaurantAddModal로 계속 전달되고 있습니다.
이는 코드 가독성을 떨어뜨리고 유지보수를 어렵게 만듭니다.
#### 불필요한 래퍼 컴포넌트:
AsideContainer와 같은 컴포넌트는 단순히 props를 전달하는 역할만 하고 있어 불필요한 중간 계층을 형성합니다.
RestaurantContainer도 마찬가지로 단순 래퍼 역할을 하고 있습니다.
#### 컴포넌트 이름의 불명확성:
RestaurantValue라는 이름은 해당 컴포넌트의 역할(레스토랑 목록 아이템)을 명확히 표현하지 못합니다.
CategoryModal은 실제로는 모달이 아닌 select 옵션을 나타내는 컴포넌트입니다.
### 개선 제안
컴포넌트 구조를 다음과 같이 단순화하는 것이 좋겠습니다:
#### 핵심 컴포넌트만 유지:
Header: 상단 헤더 컴포넌트
RestaurantList: 레스토랑 목록 및 필터링 기능
RestaurantItem: 각 레스토랑 항목 (현재 RestaurantValue)
Modal: 기본 모달 컴포넌트
RestaurantDetailModal: 레스토랑 상세 정보 모달
RestaurantAddModal: 레스토랑 추가 모달
#### 중간 래퍼 컴포넌트 제거:
AsideContainer, RestaurantContainer 등의 단순 래퍼 컴포넌트를 제거하고 직접 필요한 컴포넌트를 렌더링하는 방식으로 변경
#### 3. 상태 관리 개선:
모달 상태와 관련 로직을 더 효율적으로 관리할 수 있는 방법 고려
필요하다면 Context API나 상태 관리 라이브러리 도입 검토

gxuoo committed:
refactor: 이미지 파일 public 폴더로 이동

import 하는 방식이 아닌 public 폴더에서 정적 URL로 접근하여 이미지 Loading

gxuoo committed:
refactor: 변수 네이밍 변경

모달 상태 관리 변수 modalState를 isModalOpen 이라는 이름으로 rename 했습니다.

또한 Restaurant의 정보를 보여주는 컴포넌트인 RestaurantValue를 RestaurantListItem 으로 변경했습니다.

gxuoo committed:
refactor: module.css 개선 및 스타일 통일

module.css를 컴포넌트 별로 적용 시켰습니다. 또한 스타일이 일치되지 않았던 코드를 하나의 스타일로 일치시켰습니다.

gxuoo committed:
refactor: 이미지 파일 경로 설정

gxuoo committed:
fix: 카테고리 검증 로직 변경

기존의 alt 속성으로 비교 연산을 진행했던 것을 category key를 추가로 만들어 해당 key로 비교 연산을 진행합니다.

gxuoo committed:
refactor: 상수 네이밍 변경

upper case에 맞춰서 변경했습니다.

gxuoo committed:
refactor: 컴포넌트 네이밍 변경 및 Context API 도입

상태 관리를 위한 RestaurantContext를 추가하였고 각 컴포넌트의 역할과 기능에 따른 네이밍을 수정하였습니다.

gxuoo reviewed (commented):
3, 4단계에 달아주신 모든 리뷰를 확인하고 코드 수정 완료했습니다! 리뷰 반영이 조금 늦어진 점 죄송합니다 🥲🥲

민석님께서 제안해주신 Context API에 대해서 알아본 뒤, `RestaurantContext` 라는 Context를 만들어 기존 App.jsx에서 생성되었던 많은 상태들을 해당 Context에 정리하여 과도한 Props Drilling을 방지할 수 있었습니다.

그리고 선언되어 있는 모든 컴포넌트들의 이름을 재확인하면서 해당 컴포넌트가 수행하는 역할과 기능을 나타낼 수 있도록 변경한 뒤, 불필요한 컴포넌트는 제거한 뒤에 중요 컴포넌트만 App.jsx에서 사용할 수 있도록 변경했습니다.

이번 4단계 미션을 진행하는 과정에서 App.jsx에서 자식 컴포넌트들에게 과도한 props를 내린다고 생각은 하고 있긴 했지만, 라이브러리 도입 검토 혹은 다른 해결 방안을 찾아보는 시간을 가지지 못했던 것 같습니다. 이번 미션을 시작으로 다양한 접근을 시도해서 조금 더 깔끔하게 코드를 작성할 수 있도록 해보겠습니다. 감사합니다🙇🏻‍♂️🙇🏻‍♂️

gxuoo committed:
Merge branch 'gxuoo' into step4
